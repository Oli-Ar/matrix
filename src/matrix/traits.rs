use super::{Data, Matrix};
use std::fmt::{Display, Formatter};
use std::ops::{Index, Mul};

impl<T: Data, const N: usize> Display for Matrix<T, { N }> {
    // Prints a two dimensional representaion of the matrix
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        for row in 0..self.y {
            for col in 0..self.x {
                write!(f, "{:>2?},", self.dat[self.x * row + col])?;
            }
            writeln!(f)?;
        }
        Ok(())
    }
}

impl<T: Data, const N: usize> Index<[usize; 2]> for Matrix<T, { N }> {
    type Output = T;
    fn index(&self, index: [usize; 2]) -> &Self::Output {
        if index[0] >= self.x || index[1] >= self.y {
            // TODO: Write better error message
            panic!("Index out of bounds.");
        }
        &self.dat[index[1] * self.x + index[0]]
    }
}

// Multiplication implementation for number * matrix
impl<T, const N: usize> Mul<T> for Matrix<T, { N }>
where
    T: Data + Mul<Output = T>,
{
    type Output = Self;
    fn mul(self, rhs: T) -> Self::Output {
        let mut dat_copy = self.dat;
        for (i, e) in self.dat.iter().enumerate() {
            dat_copy[i] = rhs * *e;
        }
        Matrix {
            dat: dat_copy,
            x: self.x,
            y: self.y,
        }
    }
}
