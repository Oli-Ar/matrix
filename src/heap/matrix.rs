// TODO: This code's all kinda shit, this all needs to be fixed, make proper getters and setters as
// rename like the data method
use std::vec::Vec;

// Heap allocated generic matrix structure
#[derive(Debug, Eq, PartialEq, Clone)]
pub struct HeapMatrix<D> {
    pub(crate) buf: Vec<D>,
    pub(crate) cols: usize,
    pub(crate) rows: usize,
}
