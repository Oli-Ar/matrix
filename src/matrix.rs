pub trait Data = Into<f64> + Copy;

// The initial generic matrix structure
// All fields are private as the struct will have methods to display the matrix
#[derive(Debug, Eq, PartialEq, Clone, Copy)]
pub struct Matrix<D: Data, const M: usize, const N: usize>
where
    [D; M * N]: Sized,
{
    pub(crate) dat: [D; M * N],
}
