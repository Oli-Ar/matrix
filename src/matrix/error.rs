use super::MatrixError;
use std::fmt::{Display, Formatter};

// This pattern is exhaustive so any new errors will have to be added here for this not to error
impl Display for MatrixError {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        match *self {
            MatrixError::IncompatibleOperation(op) => {
                writeln!(f, "Incompatible operation: {}", op)?
            }
        };
        Ok(())
    }
}

// use std::error::Error;

// impl Error for MatrixError {
//     fn source(&self) -> Option<&(dyn Error + 'static)> {
//         // NOTE: Implement later if necessary
//     }
// }
