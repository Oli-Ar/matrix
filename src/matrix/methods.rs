use super::{Data, Matrix};
use std::mem::transmute;

impl<D: Data, const M: usize, const N: usize> Matrix<D, M, N>
where
    D: From<i32>,
    [D; M * N]: Sized,
{
    // Compiler will assets that passed in array is correct size so returing a result is not
    // required
    pub const fn new(data: [[D; N]; M]) -> Self {
        // This should be safe as nested arrays are guaranteed to have no padding or reordering so
        // transmuting to flatten it is safe
        // This is done so the new function is const and doesn't required a for loop to copy the
        // data into the flattened array
        let dat: &[D; M * N] = unsafe { transmute::<&[[D; N]; M], &[D; M * N]>(&data) };
        Matrix { dat: *dat }
    }

    // To create column vectors without having to pass in [[1], [2], [3], ... , [n]]
    // Doesn't return a result as the function can't fail as long as the passed arguments can be
    // passed
    pub const fn vector(dat: [D; M * N]) -> Self {
        Matrix { dat }
    }

    // For getting the raw data stored in the matrix struct
    pub const fn raw(self) -> [D; M * N] {
        self.dat
    }

    // Returns the data stored in the matrix struct in 2D array form
    // Same safety applied as when creating an array from a 2d array as seen in the new method -
    // this should be fine to do
    pub const fn data(self) -> [[D; N]; M] {
        *unsafe { transmute::<&[D; M * N], &[[D; N]; M]>(&self.dat) }
    }
}
