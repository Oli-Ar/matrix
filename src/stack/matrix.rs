// The initial generic matrix structure
// All fields are private as the struct will have methods to display the matrix
#[derive(Debug, Eq, PartialEq, Clone, Copy)]
pub struct StackMatrix<D, const M: usize, const N: usize>
where
    [D; M * N]: Sized,
{
    pub(crate) buf: [D; M * N],
}
