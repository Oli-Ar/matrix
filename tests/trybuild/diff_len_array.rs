use matrix::matrix::Matrix;

const LENGTH: usize = 9;

fn main() {
    let _ = Matrix::<i32, LENGTH>::new([[1, 2, 3], [4, 5, 6], [7]]);
}
