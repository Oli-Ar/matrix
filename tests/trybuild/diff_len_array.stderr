error[E0308]: mismatched types
 --> $DIR/diff_len_array.rs:6:63
  |
6 |     let _ = Matrix::<i32, LENGTH>::new([[1, 2, 3], [4, 5, 6], [7]]);
  |                                                               ^^^ expected an array with a fixed size of 3 elements, found one with 1 element
